runtime/secrets/reader.go

```diff
@@ -0,0 +1,192 @@
+/*
+Copyright 2025 The Flux authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package secrets
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net/url"
+
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// TLSConfigFromSecret creates a TLS configuration from a Kubernetes secret.
+//
+// The function looks for TLS certificate data in the secret using standard
+// field names (tls.crt, tls.key, ca.crt). If WithDeprecatedFieldSupport
+// option is provided, it will also check deprecated field names (certFile,
+// keyFile, caFile) as fallbacks.
+//
+// Standard field names always take precedence over deprecated ones.
+func TLSConfigFromSecret(ctx context.Context, c client.Client, name, namespace string, opts ...Option) (*tls.Config, error) {
+	options := makeOptions(opts)
+
+	secret, err := getSecret(ctx, c, name, namespace)
+	if err != nil {
+		return nil, err
+	}
+
+	tlsCert, err := getSecretData(secret, TLSCertKey, TLSCertFileKey, options.supportDeprecatedFields)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get TLS certificate: %w", err)
+	}
+
+	tlsKey, err := getSecretData(secret, TLSKeyKey, TLSKeyFileKey, options.supportDeprecatedFields)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get TLS private key: %w", err)
+	}
+
+	// CA certificate is optional, ignore error if not found
```

Comment:
Both CA and mTLS are optional, but at least one should be present.

Reply:
ref: https://github.com/fluxcd/pkg/pull/950#discussion_r2164075103